### DISCUSSION TITLE
screenshot bitmap

### DISCUSSION THREAD

**Discussion by git-bone** [mapsforge/vtm#1134] [Q&A] [answered]
Hi there,

is it possible in mapsforge vtm to create a bitmap of a mapview? i've tried with drawing on canvas and even pixelcopy. Maybe i'm doing something wrong here, but all bitmaps are blank. Use case is showing miniature bitmaps of a rendered map at a certain zoom level. There seems to be nothing out of the box in mapsforge vtm for this purpose?

**Comment by devemux86** [comment, marked as answer]
Android or desktop?

This is OpenGL, better search online how to take OpenGL screenshots from Android `GLSurfaceView` or desktop libGDX.

**Comment by demiantres** [comment]
You can use these functions:

`glPixelStorei(GL_PACK_ALIGNMENT, 1);`

`glReadPixels(0, 0, w, h, GL_RGBA, GL_UNSIGNED_BYTE, buffer_to_write_to);`

**Comment by git-bone** [comment]
yep, thanks! to both of you. I found a way that seems to work :)

edit: hm, this only seems to work when the mapView is actually visible on the screen (android).  This is the code I use in the main activity where the map currently renders in a layout. However I would like this to all happen in memory. Any pointers on how to do that? without rendering in a view, it gives me this error: IllegalArgumentException: Surface isn't valid, source.isValid() == false

`private fun getBitMapFromSurfaceView(view: MapView, callback: (Bitmap?) -> Unit) {`
`        val bitmap: Bitmap = Bitmap.createBitmap(`
`            view.width,`
`            view.height,`
`            Bitmap.Config.ARGB_8888`
`        )`
`        try {`
`            // Create a handler thread to offload the processing of the image.`
`            val handlerThread = HandlerThread("PixelCopier")`
`            handlerThread.start()`
`            PixelCopy.request(`
`                view, bitmap,`
`                { copyResult ->`
`                    if (copyResult == PixelCopy.SUCCESS) {`
`                        callback(bitmap)`
`                    }`
`                    handlerThread.quitSafely()`
`                },`
`                Handler(handlerThread.looper)`
`            )`
`        } catch (e: IllegalArgumentException) {`
`            callback(null)`
`            e.printStackTrace()`
`        }`
`    }`

**Reply by demiantres** [reply]
You can render the map into an OpenGL frame buffer object.
